# Проект по SQL
Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. 

# Цели и задачи проекта

Задача — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

# Описание данных 

**Таблица `books`**

Содержит данные о книгах:
    
`book_id` — идентификатор книги;

`author_id` — идентификатор автора;

`title` — название книги;

`num_pages` — количество страниц;

`publication_date` — дата публикации книги;

`publisher_id` — идентификатор издателя.


**Таблица `authors`**

Содержит данные об авторах:
    
`author_id` — идентификатор автора;

`author` — имя автора.



**Таблица `publishers`**

Содержит данные об издательствах:
    
`publisher_id` — идентификатор издательства;

`publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:
    
`rating_id` — идентификатор оценки;

`book_id` — идентификатор книги;

`username` — имя пользователя, оставившего оценку;

`rating` — оценка книги.



**Таблица reviews**

Содержит данные о пользовательских обзорах:
    
`review_id` — идентификатор обзора;

`book_id` — идентификатор книги;

`username` — имя автора обзора;

`text` — текст обзора.

# Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;

- Для каждой книги посчитайте количество обзоров и среднюю оценку;

- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

- Cосчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

# Как выполнить задание?

- Опишите цели исследования;

- Исследуйте таблицы — выведите первые строки;

- Сделайте по одному SQL-запросу для решения каждого задания;

- Выведите результаты каждого запроса в тетрадке;

- Опишите выводы по каждой из решённых задач.


# Примечания

Не забывайте про функции, они упрощают работу;

Получите ответы, используя только язык запросов SQL. Pandas допускается только для хранения и вывода результатов.

## Импорт библиотек,подключение к БД

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

**Добавим функцию для выгрузки таблиц**

In [3]:
def select(query):
    return pd.read_sql(query,con = engine)

**Добавим функцию для вывода информации о таблицах**

In [4]:
def inform(df):
    display(df.head()) # вывод первых 5 строк таблицы
    df.info() # получение информации
    print ('Количество дубликатов -', df.duplicated().sum()) # проверка на дубликаты
    print ('Число пропусков -',df.isna().sum())# число пропусков
    return df

## Выгрузка и изучение данных

**Таблица `books`**

In [5]:
books = '''
          SELECT * FROM books
        '''
books = select(books)
books = inform(books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
Количество дубликатов - 0
Число пропусков - book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64


Таблица `books` содержит 1000 записей в 6 столбцах.Пропусков и дубликатов не обнаружено. 

**Таблица `authors`**

In [6]:
authors = '''
            SELECT * FROM authors
          '''
authors = select(authors)
authors = inform(authors)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
Количество дубликатов - 0
Число пропусков - author_id    0
author       0
dtype: int64


Таблица `authors` содержит 636 записей в 2 столбцах.Пропусков и дубликатов не обнаружено. 

**Таблица `publishers`**

In [7]:
publishers = '''
            SELECT * FROM publishers
             '''
publishers = select(publishers)
publishers = inform(publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
Количество дубликатов - 0
Число пропусков - publisher_id    0
publisher       0
dtype: int64


Таблица `publishers` содержит 340 записей в 2 столбцах.Пропусков и дубликатов не обнаружено. 

**Таблица `ratings`**

In [8]:
ratings= ''' 
            SELECT * FROM ratings
         '''
ratings = select(ratings)
ratings = inform(ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
Количество дубликатов - 0
Число пропусков - rating_id    0
book_id      0
username     0
rating       0
dtype: int64


Таблица `ratings` содержит 6456 записей в 4 столбцах.Пропусков и дубликатов не обнаружено. 

**Таблица `reviews`**

In [9]:
reviews = '''
            SELECT * FROM reviews
          '''
reviews = select(reviews)
reviews = inform(reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
Количество дубликатов - 0
Число пропусков - review_id    0
book_id      0
username     0
text         0
dtype: int64


Таблица `reviews` содержит  2793 записи в 4 столбцах.Пропусков и дубликатов не обнаружено. 

## Задания

### Cколько книг вышло после 1 января 2000 года?

In [10]:
query = '''
            SELECT COUNT(book_id)
            FROM books 
            WHERE publication_date > '01-01-2000'
        '''
select(query)


count
0    819

После 1 января 2020 года выпущено 819 книг из пр.едставлееных 1000.

### Расчет для каждой книги  количества обзоров и среднюю оценку

In [11]:
query = '''
    SELECT 
        books.book_id,
        books.title,
        COUNT(DISTINCT reviews.text) AS count_reviews,
        ROUND(AVG(ratings.rating), 2) AS avg_rating
    FROM  books
    LEFT JOIN ratings ON ratings.book_id = books.book_id
    LEFT JOIN reviews ON reviews.book_id = books.book_id
    GROUP BY 
         books.title,
         books.book_id
    ORDER BY
    count_reviews DESC,
    avg_rating DESC
 '''
select(query)    

book_id                                              title  \
0        948                            Twilight (Twilight  #1)   
1        302  Harry Potter and the Prisoner of Azkaban (Harr...   
2        299  Harry Potter and the Chamber of Secrets (Harry...   
3        656                                     The Book Thief   
4        734                                   The Glass Castle   
..       ...                                                ...   
995      191  Disney's Beauty and the Beast (A Little Golden...   
996      387                               Leonardo's Notebooks   
997      221                          Essential Tales and Poems   
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
999      808                            The Natural Way to Draw   

     count_reviews  avg_rating  
0                7        3.66  
1                6        4.41  
2                6        4.29  
3                6        4.26  
4                6        4.21  
..             ...         ...  
995              0        4.00  
996              0        4.00  
997              0        4.00  
998              0        3.67  
999              0        3.00  

[1000 rows x 4 columns]

Посмотрим,что из себя прадставляет шкала оценок

In [12]:
query = '''
    SELECT 
      DISTINCT rating
    FROM  ratings
 '''
select(query)  

rating
0       1
1       3
2       5
3       2
4       4

Расчитали для каждой книги количество обзоров и среднюю оценку пользователей. Больше всего обзоров  у книги `Twilight`('Сумерки')- **7** при средней оценке пользователей **3.66**.Однако есть книги и без обзоров.Максимальный возможный рейтинг **5** баллов.Высокий рейтинг **4.41** и **4.29** при среднем числе обзоров -**6** имеют книги о Гарри Поттере

### Определение издательства, которое выпустило наибольшее число книг толще 50 страниц

In [13]:
query = '''
            SELECT
                   publishers.publisher,
                   COUNT(books.book_id) AS count_books      
            FROM books           
            LEFT JOIN publishers AS publishers ON books.publisher_id=publishers.publisher_id
            WHERE books.num_pages > 50
            GROUP BY books.publisher_id, publishers.publisher_id
            ORDER BY count_books DESC
            LIMIT 1            
        '''
select(query) 

publisher  count_books
0  Penguin Books           42

Издательство, которое выпустило наибольшее число книг толще 50 страниц  - это `Penguin Books`(42 книги).Это известное крупное британское издательство.

### Определение автора с самой высокой средней оценкой книг(учитывайте только книги с 50 и более оценками)

In [14]:
query = '''

     SELECT
          authors.author,
          ROUND(AVG(ratings.rating),2) AS avg_rating
    FROM ratings
    JOIN books ON books.book_id = ratings.book_id
    JOIN authors ON books.author_id = authors.author_id
    WHERE ratings.book_id IN (
          SELECT ratings.book_id
          FROM ratings
          GROUP BY ratings.book_id
          HAVING COUNT(ratings.book_id) >= 50)
    GROUP BY authors.author_id
    ORDER BY AVG(ratings.rating) DESC
    LIMIT 5
    '''
select(query) 

author  avg_rating
0         J.K. Rowling/Mary GrandPré        4.29
1  Markus Zusak/Cao Xuân Việt Khương        4.26
2                     J.R.R. Tolkien        4.25
3                  Louisa May Alcott        4.19
4                       Rick Riordan        4.08

Автор с максимальной рейтинговой оценкой - `J.K. Rowling/Mary GrandPré`(Джоан К. Роулинг/Мэри Гранпре).Оценка - **4.29** Это автор всемирно известной серии книг о Гарри Поттере

### Расчет среднего количество обзоров от пользователей, которые поставили больше 48 оценок.

In [15]:
query = '''
        SELECT ROUND(AVG(r.count),2) AS avg_review
        FROM 
            (SELECT count(*)
             FROM reviews
             WHERE username IN
             (SELECT username
             FROM ratings
             GROUP BY username
             HAVING count(rating) > 48)
        GROUP BY username) AS r
'''
select(query) 

avg_review
0        24.0

Среднее  количество обзоров от пользователей, которые поставили больше 48 оценок - **24** обзоров

## Вывод

**Задача проекта**  — проанализировать базу данных с  информацией о книгах, издательствах, авторах, а также пользовательские обзоры книг.Получить данные, которые  данные помогут сформулировать ценностное предложение для нового продукта(компания решила быть на волне и купила крупный сервис для чтения книг по подписке.)

Для анализа были  выгружены  таблицы :

- `authors` содержащую  636 записей в 2 столбцах о авторах книг

- `books` содержащую 1000 записей в 6 столбцах о названии книг,количестве страниц в них,дате выпуска             и идентефикаторах издателя,книги и автора
 
- `publishers` содержащую  340 записей в 2 столбцах о издательствах

- `ratings` содержащую 6456 записей в 4 столбцах о пользовательских оценках книг

- `reviews` содержащую 2793 записи в 4 столбцах о отзывах пользователей

Пропусков и дубликатов в них не обнаружено.

**Получены ответы на вопросы задания:**

**Посчитайте, сколько книг вышло после 1 января 2000 года**
   
   - После 1 января 2020 года выпущено 819 книг из представлееных 1000. 

**Для каждой книги посчитайте количество обзоров и среднюю оценку**
   
   - Расчитали для каждой книги количество обзоров и среднюю оценку пользователей. Больше всего           обзоров у книги Twilight('Сумерки')- 7 при средней оценке пользователей 3.66.Однако есть книги и       без обзоров.Максимальный возможный рейтинг 5 баллов.Высокий рейтинг 4.41 и 4.29 при среднем           числе  обзоров -6 имеют книги о Гарри Поттере
   
**Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюр**
   
   - Издательство, которое выпустило наибольшее число книг толще 50 страниц  - это `Penguin Books`(42      книги).Это известное крупное британское издательство.
   
**Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более          оценками**
   
   - Автор с максимальной рейтинговой оценкой - `J.K. Rowling/Mary GrandPré`(Джоан К. Роулинг/Мэри         Гранпре).Оценка - **4.29** Это автор всемирно известной серии книг о Гарри Поттере
   
**Cосчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.**
  
   - Среднее  количество обзоров от пользователей, которые поставили больше 48 оценок - **24** обзоров
  
В целом можно сказать, что более востребованна литература жанра фентези.

  **Рекомендации**
  
  - Стоит ориентироваться на новые изданные книги.
  
  - Предлагать книги с высокими пользовательскими оценкми и учетом отзывов
  
  - Предлагать книги популярных писателей и издательств
  
  - Сделать акцент на легких жанрах.
  
  